### PR TITLE
Release 1.13.0

### LINKED ISSUE #4246: Create a non-Meteor factory/fixtures solution
## Goal
Remove the large similar objects we've been copy-pasting in all of the integration tests (and some Jest unit tests). Instead use a factory pattern to get fake data.

## Proposal
Here is my proposal for API, but feel free to alter this as necessary to get everything we need to keep our tests simple.

Make it an NPM package (maybe `@reactioncommerce/data-factory`?)
Depend on https://www.npmjs.com/package/simpl-schema-mockdoc

In the reaction files where schemas are defined, do something like:

```js
import { createFactoryForSchema } from "@reactioncommerce/data-factory";
createFactoryForSchema("Shop", ShopSchema);
```

This will add `Shop` property to a `Factory` object that can be imported in tests, in devtools, anywhere.

```js
import { Factory } from "@reactioncommerce/data-factory";

const shop = Factory.Shop.makeOne({ prop: val });
const shops = Factory.Shop.makeMany(5, { prop: val });
```

`makeOne` returns a single object to be inserted/used, while `makeMany` returns multiple in an array (array length = first arg value).

The `{ prop: val }` object is where you specify override properties you specifically want. For any properties not there, or if you omit that object, you'll get back the object created by the `simpl-schema-mockdoc` package.

If `simpl-schema-mockdoc` does not meet our needs or is buggy, we can copy the relevant code into this data-factory package and update it to work for us.

## Work
1. Create the package
1. Add the package to Reaction, pass in all the schemas.
1. Change all of the integration tests in the `/tests` folder to use this to build their mock data objects.

### LINKED ISSUE #4263: Make SimpleSchema definitions work without Meteor
## Work

Move/change the lib/collections/schemas files so that we can get access to schemas in devserver/Jest tests without a Meteor context.

### LINKED ISSUE #4326: Importer errors with Products causes app to crash with an uncaught error
## Issue Description

Don't import data into Products, Shipping, and Tags collection if data already exists. The importer sometimes freaks out and tries to re-import the same data into those collections with the same ID and it causes to the app to crash on startup.

(This error is from an added console log of the uncaught error in the importer)
```
reaction_1   | Error: WriteError({"code":11000,"index":0,"errmsg":"E11000 duplicate key error collection: reaction.Products index: _id_ dup key: { : \"BCTMZ6HTxFSppJESk\" }","op":{"q":{"title":"Basic Reaction Product","_id":"BCTMZ6HTxFSppJESk"},"u":{"$set":{"_id":"BCTMZ6HTxFSppJESk","shopId":"J8Bhq3uTtdgwZx3rz","ancestors":[],"description":"Sign in as administrator to edit.\nYou can clone this product from the product grid.\nYou can upload images click or drag in image box on the left here.\nTag this product below, and then add tag in navigation.\nClick the bookmark in the tag to set product url.\nOption variants, price, quantity, and child variants are created by clicking on the variant below, clone the variant to add more options.\nDetails can be added below the image for more specific product information.\n Login next to the cart, and then click the dashboard icon for more tools.","handle":"basic-reaction-product","hashtags":["rpjCvTBGjhBi2xdro","cseCBSSrJ3t8HQSNP"],"price.range":"12.99 - 19.99","price.min":12.99,"price.max":19.99,"isVisible":true,"isLowQuantity":false,"isSoldOut":false,"isBackorder":false,"metafields":[{"key":"Material","value":"Cotton"},{"key":"Quality","value":"Excellent"}],"pageTitle":"This is a basic product. You can do a lot with it.","type":"simple","title":"Basic Reaction Product","updatedAt":"2018-06-14T16:50:16.381Z","vendor":"Example Manufacturer"},"$setOnInsert":{"createdAt":"2018-06-14T16:50:16.380Z","requiresShipping":true,"isDeleted":false,"template":"productDetailSimple","workflow":{"status":"new"}}},"multi":true,"upsert":true}})
reaction_1   |     at buffer.execute (server/api/core/importer.js:130:26)
reaction_1   |     at err (/home/node/.meteor/packages/npm-mongo/.3.0.7.g0jse4.0t1w++os+web.browser+web.browser.legacy+web.cordova/npm/node_modules/mongodb/lib/utils.js:415:14)
reaction_1   |     at session.endSession (/home/node/.meteor/packages/npm-mongo/.3.0.7.g0jse4.0t1w++os+web.browser+web.browser.legacy+web.cordova/npm/node_modules/mongodb/lib/utils.js:399:27)
reaction_1   |     at ClientSession.endSession (/home/node/.meteor/packages/npm-mongo/.3.0.7.g0jse4.0t1w++os+web.browser+web.browser.legacy+web.cordova/npm/node_modules/mongodb-core/lib/sessions.js:72:41)
reaction_1   |     at executeCallback (/home/node/.meteor/packages/npm-mongo/.3.0.7.g0jse4.0t1w++os+web.browser+web.browser.legacy+web.cordova/npm/node_modules/mongodb/lib/utils.js:397:17)
reaction_1   |     at handleCallback (/home/node/.meteor/packages/npm-mongo/.3.0.7.g0jse4.0t1w++os+web.browser+web.browser.legacy+web.cordova/npm/node_modules/mongodb/lib/utils.js:128:55)
reaction_1   |     at /home/node/.meteor/packages/npm-mongo/.3.0.7.g0jse4.0t1w++os+web.browser+web.browser.legacy+web.cordova/npm/node_modules/mongodb/lib/bulk/unordered.js:528:20
reaction_1   |     at handleCallback (/home/node/.meteor/packages/npm-mongo/.3.0.7.g0jse4.0t1w++os+web.browser+web.browser.legacy+web.cordova/npm/node_modules/mongodb/lib/utils.js:128:55)
reaction_1   |     at resultHandler (/home/node/.meteor/packages/npm-mongo/.3.0.7.g0jse4.0t1w++os+web.browser+web.browser.legacy+web.cordova/npm/node_modules/mongodb/lib/bulk/unordered.js:451:5)
reaction_1   |     at /home/node/.meteor/packages/npm-mongo/.3.0.7.g0jse4.0t1w++os+web.browser+web.browser.legacy+web.cordova/npm/node_modules/mongodb-core/lib/connection/pool.js:544:18
reaction_1   |     at _combinedTickCallback (internal/process/next_tick.js:131:7)
reaction_1   |     at process._tickDomainCallback (internal/process/next_tick.js:218:9)
```

## Steps to Reproduce

1. With a fresh app go to the PDP page as an admin
2. Change the product title. I did `Basic Reaction Product--`
3. Restart the app either with a code change or, stop/start
4. Observe the app not starting due to an error.

## Possible Solution

- `count === 0` to wrap import statements in `/server/startup/load-data.js`
- Guard against errors in `/server/api/core/importer.js` in `Importer.commit`

### Versions

Node: 9.10.1
NPM: 6.1.0
Meteor Node: 8.11.2
Meteor NPM: 5.10.0
Reaction CLI: 0.29.0
Reaction: 1.13.0
Reaction branch: feat-4319-mikemurray-pdp-catalog
Docker: 18.03.1-ce

### PR DESCRIPTION
**Pull request by spencern**
# v1.13.0

## Removal of Legacy Product Revision Control system
The major change in this release is that we've removed the existing revision control system in favor of publishing Products to the Catalog. The existing revision control system contained some powerful ideas, but was complex and intertwined into many areas of the app that were not directly related to Product. By removing the old revision control system, we've paved the way to substantially improve product grid performance, especially for Operators and we've simplified the product publication logic.

This removes all code, hooks, collections, and packages related to revision control.

**Breaking changes:**
 - Any custom packages that depend on the `Revision` control system.
 - Operators will no longer be able to "undo" changes to a product. Instead products are published through the catalog
 - Any _unpublished_ changes to products will be lost when upgrading to `1.13.0`. In this release, it's possible that if you were to downgrade that you would see the unpublished changes again, but that may cause unexpected behavior. We recommend either publishing or discarding any changes to your products before upgrading to this release.
 - Similarly, any products that have been created but not published will demonstrate unexpected behavior. To avoid this undesirable behavior, publish any newly created, unpublished products prior to upgrading to this release.

 - **All plugin authors** will need to update your `package.json` with a change to the babel config similar to what was done [here](https://github.com/reaction-contrib/meteor-authorize-net/commit/f19a5cf7591a17f426e67bd3737af5a4d1c7a64a)

## Update to Meteor 1.7
This update brings some enormous improvements to the amount of time it takes to rebuild the application in development after making a file change. In some (less than perfectly scientific) tests that I ran testing file changes between 1.7 and 1.6.1 I saw 50%-90% improvements in the reload time. Your experience may vary depending on how much you've customized Reaction, your computer specs, and your specific development setup, but I fully expect this to be a noticeable improvement for anyone working with Reaction.

There are some [early](https://github.com/meteor/meteor/issues/9949) [reports](https://github.com/meteor/meteor/issues/9945) that the included update to the MongoDB driver may have [some kinks](https://github.com/meteor/meteor/issues/9944) to work out, so I'd follow those issues on Meteor's repo if that's a cause for concern for you. These reports are all coming from a single person, and we haven't experienced any of these issues in particular yet, but we'll be keeping an eye on them.

##
We ran into a few issues with `npm install` that we resolved in #4317. One product that came out of this investigation was some documentation for how to properly clean up and rebuild docker images in Reaction.

To stop and clean up your images
```sh
  docker-compose down -v --rmi local --remove-orphans
```

To rebuild your Reaction images
```sh
  docker-compose up --build --force-recreate --renew-anon-volumes
```

If you only want to run Reaction and not the GraphQL DevServer
```sh
  docker-compose up --build --force-recreate --renew-anon-volumes reaction
```

If you only want to run the DevServer and not the Meteor app
```sh
  docker-compose up --build --force-recreate --renew-anon-volumes devserver
```

### Performance
 - perf: remove revision control (#4238)
 - perf: update to Meteor 1.7 (#4265)

### Bug Fixes
 - fix: Use catalog collection for PDP (#4324)
 - fix: Import fixture data only if collections empty (#4327) .. Resolves #4326
 - fix: Invalid class name: .variant-list-item-{variant._id} (#4217)
 - fix: NPM build issue in Docker build (#4317)
 - fix: add getAutoValues: false to discounts/codes/remove (#4288)
 - fix: hadolint image version (#4306)
 - fix: CI step failure to tag Docker image with latest release version (#4304)
 - fix: admin products publication slowness (#4260)
 - fix: remove inventoryPolicy check on low inventory (#4298)

### Refactors
 - refactor: non meteor schemas (#4266) .. Resolves #4263

### Tests
 - test: new mocks factory (#4276) .. Resolves #4246
 - test: run snyk when package.json has changed or base is master (#4285)

### Docs
 - docs(jsdoc): document all Meteor Template helpers in 1 @namespace (#3841) .. Resolves #3840

### Features
 - feat: add Media to Tag Schema and GraphQL query (#4270)

### Chore
 - chore: update default GraphQL query limit values (#4297)

## Contributors
Thanks to @mikeumus for contributing to this release. 🎉

### CODE REVIEW

**Comment by dhonig** [comment]
Revision control is a great and powerful feature.  Its also not required by most e-commerce sites so glad to see this in motion!  :+1:

File: imports/plugins/core/catalog/client/containers/publishContainer.js [hunk 1]
```diff
@@ -0,0 +1,83 @@
+import React, { Component } from "react";
+import PropTypes from "prop-types";
+import { composeWithTracker } from "@reactioncommerce/reaction-components";
+import PublishControls from "../components/publishControls";
+import { Meteor } from "meteor/meteor";
```
Comment (zenweasel) [inline imports/plugins/core/catalog/client/containers/publishContainer.js]: Import order

Comment (spencern) [reply imports/plugins/core/catalog/client/containers/publishContainer.js]: Why isn't eslint catching this?

File: server/publications/collections/products.js [hunk 2]
```diff
@@ -165,33 +128,48 @@ function filterProducts(productFilters) {
...
+
+  return selector;
+}
+
+function filterProducts(productFilters) {
+  // if there are filter/params that don't match the schema
+  // validate, catch except but return no results
+  try {
+    if (productFilters) filters.validate(productFilters);
+  } catch (e) {
```
Comment (zenweasel) [inline server/publications/collections/products.js]: Do we make an exception re: single letter variables here?

Comment (spencern) [reply server/publications/collections/products.js]: shouldn't be, looks like it slipped through code review. Nice catch.

Comment (spencern) [reply server/publications/collections/products.js]: Looks like it came in through a community contribution, still shouldn't have slipped through code review though. I'll fix it.

Comment (zenweasel) [reply server/publications/collections/products.js]: I would have swore that ESLint caught those in the past

**Pull request merged**